File: amzn/Linked Lists/MergekSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if(lists.length ===0 ) {
        return null;
    }
    while(lists.length > 1){
        let a = lists.shift(); 
        let b = lists.shift(); 
        let mergedList = mergeTwoLists(a,b);
        lists.push(mergedList);
    }
    return lists[0];
}  
function mergeTwoLists(a,b){
    let tmpHead = new ListNode();
    let curr = tmpHead;
    
    while(a!==null&&b!==null){
        if(a.val < b.val){
            curr.next = a;
            a = a.next;
        }else{
            curr.next = b;
            b = b.next;
        }
        curr = curr.next;
    }
    if(a!==null){
        curr.next = a   
    }
     if(b!==null){
        curr.next = b  
    }
    return tmpHead.next;
}

/***********RECURSIVE*************/
var mergeKLists = function(lists) {
    let merged = null;
    for(list of lists){
        merged = merge(merged,list);
    }
    return merged;
};

var merge = function(l1, l2){
     if (l2 && l1) {
        if (l1.val < l2.val) {
          l1.next = merge(l1.next, l2);
          return l1;
        }
        l2.next = merge(l2.next, l1);
        return l2;
    }
    return l1 || l2;
}

File: amzn/Linked Lists/golang/MergekSortedLists.go
/**
Merge with Divide And Conquer
Intuition & Algorithm

This approach walks alongside the one above but is improved a lot. 
We don't need to traverse most nodes many times repeatedly

Pair up \text{k}k lists and merge each pair.

After the first pairing, \text{k}k lists are merged into k/2k/2 
lists with average 2N/k2N/k length, then k/4k/4, k/8k/8 and so on.

Repeat this procedure until we get the final sorted linked list.

Thus, we'll traverse almost NN nodes per pairing and merging, and repeat 
this procedure about \log_{2}{k}log 2k times.
**/
 
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
	listsLen := len(lists)
	if listsLen == 0 {
		return nil
	} else if listsLen == 1 {
		return lists[0]
	}

	for len(lists) > 1 {
		list1, list2 := lists[0], lists[1]
		merged := mergeLists(list1, list2)

		lists = lists[2:]
		lists = append(lists, merged)
	}
	return lists[0]
}

func mergeLists(listA, listB *ListNode) *ListNode {
	res := &ListNode{}
	dummyHead := res
	for listA != nil || listB != nil {
		if listA == nil {
			res.Next = listB
			break
		}
		if listB == nil {
			res.Next = listA
			break
		}

		if listA.Val < listB.Val {
			res.Next = listA
			res = res.Next
			listA = listA.Next
		} else {
			res.Next = listB
			res = res.Next
			listB = listB.Next
		}
	}
	return dummyHead.Next
}
